### PR TITLE
Passed the scheduling argument through the `*_generator` function.

### PR DESCRIPTION
**Pull request by joeyearsley**
Previously, the scheduling parameter was not passed to the OrderedEnqueuer, which was needed in training.
Since the scheduling parameter is currently a binary decision, I've also altered the value to a boolean and changed the kwarg to be more descriptive.

It also, seems we may need a shuffle function inside the Sequence call. Currently, only the input is shuffled, whilst this ensures some randomness this could be improved on via a function call.

Also, @fchollet do you remember why in the generators you passed different seeds? - This is also missing from the current code, however, it is trivial to implement.

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -1932,7 +1940,8 @@ def evaluate_generator(self, generator, steps,
         try:
             if is_sequence:
                 enqueuer = OrderedEnqueuer(generator,
-                                           use_multiprocessing=use_multiprocessing)
+                                           use_multiprocessing=use_multiprocessing,
+                                           ordered=ordered)
```
Comment (Dref360) [inline keras/engine/training.py]: Does it make sense?  evalute_generator(...,ordered=True) ==evalute_generator(...,ordered=False)

File: keras/engine/training.py [hunk 2]
```diff
@@ -2015,6 +2025,8 @@ def predict_generator(self, generator, steps,
                 non picklable arguments to the generator
                 as they can't be passed
                 easily to children processes.
+            ordered: Sequential querying of data if `True`,
+                random otherwise.
```
Comment (Dref360) [inline keras/engine/training.py]: What's the point of shuffling for predict?

File: keras/utils/data_utils.py [hunk 3]
```diff
@@ -434,15 +434,15 @@ class OrderedEnqueuer(SequenceEnqueuer):
...
     """
 
     def __init__(self, sequence,
                  use_multiprocessing=False,
-                 scheduling='sequential'):
+                 ordered=True):
         self.sequence = sequence
         self.use_multiprocessing = use_multiprocessing
-        self.scheduling = scheduling
+        self.ordered = ordered
```
Comment (Dref360) [inline keras/utils/data_utils.py]: Should we test that ordered work? Probably.

File: keras/engine/training.py [hunk 4]
```diff
@@ -1680,6 +1681,8 @@ def fit_generator(self, generator,
                 non picklable arguments to the generator
                 as they can't be passed
                 easily to children processes.
+            ordered: Sequential querying of data if `True`,
```
Comment (Dref360) [inline keras/engine/training.py]: Only for Sequence. No effect for Generators

File: keras/utils/data_utils.py (hunk 3 above)
Comment (joeyearsley) [inline keras/utils/data_utils.py]: What do you mean test it worked?

File: keras/engine/training.py (hunk 2 above)
Comment (joeyearsley) [inline keras/engine/training.py]: That was mostly to keep kwargs consistent.

File: keras/engine/training.py (hunk 1 above)
Comment (joeyearsley) [inline keras/engine/training.py]: Same as the predict, to keep kwargs consistent.

File: keras/utils/data_utils.py (hunk 3 above)
Comment (Dref360) [inline keras/utils/data_utils.py]: Just a test that would validate that if ordered=False, there is some shuffling. Nothing fancy, but I could prevent some bugs in the future.

Comment (joeyearsley) [inline keras/utils/data_utils.py]: Sure, I'll add it in, good spot.

**Comment by fchollet** [comment]
We already have a `shuffle` argument in `fit`. Is this the same thing?

**Comment by Dref360** [comment]
Pretty much, but for Sequence. So every epoch the order of the batches will be shuffled.

**Comment by fchollet** [comment]
1 - if this only affects `Sequence`, shouldn't it be part of its API, and not part of the `*_generator` API?
2 - the API should be consistent across the codebase, so we should use `shuffle`, not `ordered`

**Comment by Dref360** [comment]
So a property in Sequence that the Enqueuer can poke?

**Comment by joeyearsley** [comment]
1 - had to pass through the *_generator methods, since the OrderedEnqueuer isn't accessible else where, unless we make users pass the enqueuer into these methods? - the plus is that it will be even more flexible, down side is more work for the user.

2 - I'll alter when I'm back from vacation

**Comment by joeyearsley** [comment]
@fchollet It keeps timing out on this test: `tests/keras/wrappers/scikit_learn_test.py::test_regression_class_build_fn`

However, I haven't touched anything related to it. How can I fix it?

**Comment by joeyearsley** [comment]
@Dref360 Thanks for the suggestion!

File: keras/engine/training.py [hunk 5]
```diff
@@ -1680,6 +1681,9 @@ def fit_generator(self, generator,
                 non picklable arguments to the generator
                 as they can't be passed
                 easily to children processes.
+            shuffle: whether to shuffle the data at the beginning of each
+                epoch. Only used with instances of Sequence (
```
Comment (fchollet) [inline keras/engine/training.py]: Use code markers around code keywords (`)

File: keras/utils/data_utils.py [hunk 6]
```diff
@@ -351,6 +351,12 @@ def __len__(self):
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def on_epoch_end(self):
+        """A function which is called at the end of the epoch.
```
Comment (fchollet) [inline keras/utils/data_utils.py]: "Method called at the end of every epoch."

File: keras/utils/data_utils.py [hunk 7]
```diff
@@ -351,6 +351,12 @@ def __len__(self):
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def on_epoch_end(self):
```
Comment (fchollet) [inline keras/utils/data_utils.py]: The addition of this method appears unrelated to the `shuffle` arg?

Comment (joeyearsley) [inline keras/utils/data_utils.py]: It is and it isn't.
The `shuffle` arg relates to the shuffling of the batch indices in the sequences, however, you may also wish to implement file path shuffling in the Sequence method. This would allow that by placing that shuffle operator into the on_epoch_end method.
The path shuffling would ensure that the files inside the batches differed as well as the batch ordering.
